### PR TITLE
Modify ICompilationLibrariesProvider to return a sequence of reference paths

### LINKED ISSUE #4738: Modify ICompilationLibrariesProvider to return a sequence of reference path strings
Based on offline discussions with @davidfowl  and @rynowak , the plan is to modify `ICompilationLibrariesProvider` to not expose `CompilationLibrary` instances instead returning a list of paths to reference assemblies. 

An additional follow up was to return the assembly path in the event the assembly was compiled without `preserveCompilationContext`. This will unblock csproj based class library scenarios which do not have deps record.

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes #4738

### CODE REVIEW

**Comment by pranavkm** [comment]
cc @davidfowl \ @rynowak

**Comment by javiercn** [comment]
:shipit:

File: src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/AssemblyPart.cs [hunk 1]
```diff
@@ -41,15 +42,15 @@ public AssemblyPart(Assembly assembly)
...
         {
             var dependencyContext = DependencyContext.Load(Assembly);
             if (dependencyContext != null)
             {
-                return dependencyContext.CompileLibraries;
+                return dependencyContext.CompileLibraries.SelectMany(library => library.ResolveReferencePaths());
             }
 
-            return new CompilationLibrary[0];
+            return new[] { Assembly.Location };
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/AssemblyPart.cs]: This is probably worth a command to explain why and what.

**Comment by rynowak** [comment]
:shipit:

File: src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/AssemblyPart.cs (hunk 1 above)
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/AssemblyPart.cs]: Did you mean a comment?

Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/AssemblyPart.cs]: I command you to add a comment

Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/AssemblyPart.cs]: > sudo add comment

File: src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/AssemblyPart.cs [hunk 2]
```diff
@@ -41,15 +41,37 @@ public AssemblyPart(Assembly assembly)
...
+                {
+                    for (var i = 0; i < compileLibraries.Count; i++)
+                    {
+                        var compileLibrary = compileLibraries[i];
+                        referencePaths.AddRange(compileLibrary.ResolveReferencePaths());
+                    }
+
+                    return referencePaths;
+                }
+                catch
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/AssemblyPart.cs]: cc @dougbu

Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/AssemblyPart.cs]: @rynowak \ @javiercn we're adding an additional guard against `ResolveReferencePaths` failing. This happened in @dougbu's scenario for https://github.com/aspnet/Mvc/issues/4497 where the assembly was compiled with `preserveCompilationContext` but is dynamically loaded at runtime.

Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/AssemblyPart.cs]: This is problematic. Having a try/catch here means that a part is *all-or-nothing* with it's dependencies. 

What's the issue that causes `ResolveReferencePaths` to throw? Would a user rather see this exception?

Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/AssemblyPart.cs]: Still seem to hit issues even w/o `preserveCompilationContext: true`. See a FNF exception with `Message` "Could not load file or assembly 'ClassLibrary, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null' or one of its dependencies. The system cannot find the file specified." Oddly the exception is `thrown` when `DefaultRoslynCompilationService` asks the successfully-generated for its exported types. Seems like the compilation dependencies (and their full paths) aren't preserved for runtime.

Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/AssemblyPart.cs]: @rynowak I don't think there's much a user would be to able to do with the exception. I'll file a follow up on the CLI to figure out guidance on dealing with this.

Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/AssemblyPart.cs]: >  I don't think there's much a user would be to able to do with the exception.

But doesn't this indicate a deployment/publishing problem? We could possibly provide a better error

Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/AssemblyPart.cs]: Spoke to @pakrym - he says as long as the dependencies are in the application base path (bin dir), `ResolveReferencePaths` should behave correctly.

Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/AssemblyPart.cs]: That only solves half the problem. What about packages referenced from the loaded assembly but not from the site?

Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/AssemblyPart.cs]: They're going to address those two. @davidfowl and @pakrym were discussing a plan for it. @pakrym could you update #4497 with the CLI issue tracking your change?

**Pull request closed without merging**